### PR TITLE
Add more utility methods to Buffer

### PR DESCRIPTION
**Pull request by NiteshKant**
__Motivation__

During the process of migrating codecs to use `Buffer`, I stumbled upon a few utility methods that may be useful to be added to `Buffer`

__Modification__

- Added `[read, write]Split()` methods to split at an offset from the current reader/writer offset.
- Added `accumulate[Reader,Writer]Offset()` methods to increment/decrement reader/writer offsets.
- Added `[read, write]CharSequence()` methods which is used extensively in HTTP.

__Result__

More utility methods to be used by codecs.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/api/Buffer.java [hunk 1]
```diff
@@ -117,6 +118,19 @@
...
+    /**
+     * Set the reader offset to {@code readerOffset() + delta}.
+     *
+     * @param delta to accumulate.
+     * @return This Buffer.
+     * @throws IndexOutOfBoundsException if the new reader offset is less than zero or greater than the current
+     * {@link #writerOffset()}.
+     * @throws BufferClosedException if this buffer is closed.
+     */
+    default Buffer accumulateReaderOffset(int delta) {
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/api/Buffer.java:130]: This name is a bit of a mouthful. How about `skipReadable` instead?

File: buffer/src/main/java/io/netty/buffer/api/Buffer.java [hunk 2]
```diff
@@ -135,6 +149,20 @@
...
+     * Set the write offset to {@code writerOffset() + delta}.
+     *
+     * @param delta to accumulate.
+     * @return This Buffer.
+     * @throws IndexOutOfBoundsException if the new writer offset is less than the current {@link #readerOffset()} or
+     * greater than {@link #capacity()}.
+     * @throws BufferClosedException if this buffer is closed.
+     * @throws BufferReadOnlyException if this buffer is {@linkplain #readOnly() read-only}.
+     */
+    default Buffer accumulateWriterOffset(int delta) {
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/api/Buffer.java:162]: Likewise `skipWritable`.

File: buffer/src/main/java/io/netty/buffer/api/Buffer.java [hunk 3]
```diff
@@ -482,6 +537,108 @@ default Buffer copy() {
      */
     Buffer copy(int offset, int length);
 
+    /**
+     * Splits the buffer into two, at the {@code offset} from current {@linkplain #readerOffset()} reader offset}
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/api/Buffer.java:541]:
```suggestion
     * Splits the buffer into two, at the {@code offset} from the current {@linkplain #readerOffset()} reader offset}
```

File: buffer/src/main/java/io/netty/buffer/api/Buffer.java [hunk 4]
```diff
@@ -482,6 +537,108 @@ default Buffer copy() {
...
+     *
+     * @return A new buffer with independent and exclusive ownership over the previously read and readable bytes from
+     * this buffer.
+     */
+    default Buffer readSplit(int offset) {
+        return split(readerOffset() + offset);
+    }
+
+    /**
+     * Splits the buffer into two, at the {@code offset} from current {@linkplain #writerOffset()} writer offset}
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/api/Buffer.java:592]:
```suggestion
     * Splits the buffer into two, at the {@code offset} from the current {@linkplain #writerOffset()} writer offset}
```

File: buffer/src/main/java/io/netty/buffer/api/BufferUtils.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+import static io.netty.util.CharsetUtil.US_ASCII;
+
+/**
+ * Utilities for buffers.
+ */
+final class BufferUtils {
+    private BufferUtils() {
+        // no instances.
+    }
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/api/BufferUtils.java:30]: These could also be put in `Statics` which is already a bit of a kitchen sink.

File: buffer/src/main/java/io/netty/buffer/api/BufferUtils.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        return new String(data, 0, length, charset);
+    }
+
+    static void writeCharSequence(CharSequence source, Buffer destination, Charset charset) {
+        if (US_ASCII.equals(charset) && source instanceof AsciiString) {
+            AsciiString asciiString = (AsciiString) source;
+            destination.writeBytes(asciiString.array(), asciiString.arrayOffset(), source.length());
+            return;
+        }
+        // TODO: Copy optimized writes from ByteBufUtil
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/api/BufferUtils.java:48]: I think it's fine for this method to be the slow fallback. Some optimisations only work for certain internal buffer representations, and those can be done in those buffer implementations.

File: buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+public class BufferCharSequenceTest extends BufferTestSupport {
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    void readCharSequence(Fixture fixture) {
+        try (BufferAllocator allocator = fixture.createAllocator()) {
+            try (Buffer buf = allocator.allocate(32)) {
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java:36]: Nit: try-with-resources statements can be combined to get less nesting. This seems to apply to all the tests added in this PR.

File: buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                String data = "Hello World";
+                buf.writeBytes(data.getBytes());
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java:38]: Let's avoid relying on the default platform encoding.

File: buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(0, buf.readerOffset());
+
+                final CharSequence read = buf.readCharSequence(data.length(), US_ASCII);
+                Assertions.assertEquals(data, read);
+                assertEquals(data.length(), buf.writerOffset());
+                assertEquals(data.length(), buf.readerOffset());
+            }
+        }
+    }
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java:84]: We could also parameterise over a number of different encodings and `CharSequence` implementations.

File: buffer/src/test/java/io/netty/buffer/api/tests/BufferOffsetTest.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ * the License.
+ */
+package io.netty.buffer.api.tests;
+
+import io.netty.buffer.api.Buffer;
+import io.netty.buffer.api.BufferAllocator;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class BufferOffsetTest extends BufferTestSupport {
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/api/tests/BufferOffsetTest.java:22]: There is already `BufferOffsetsTest`.

**Review by chrisvest — CHANGES_REQUESTED** [review]

File: buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java (hunk 9 above)
Comment (NiteshKant) [reply buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java:84]: In the interest of time and the fact that we are depending on the contract of `CharSequence`, lemme punt that for now

File: buffer/src/main/java/io/netty/buffer/api/BufferUtils.java (hunk 6 above)
Comment (NiteshKant) [reply buffer/src/main/java/io/netty/buffer/api/BufferUtils.java]: When I have the HTTP changes in place, let us discuss in the context of those which seems to be the reason why the optimizations were added to `ByteBufUtil`.

File: buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import java.nio.charset.StandardCharsets;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+public class BufferCharSequenceTest extends BufferTestSupport {
+
+    @ParameterizedTest
+    @MethodSource("allocators")
+    void readCharSequence(Fixture fixture) {
+        try (BufferAllocator allocator = fixture.createAllocator(); Buffer buf = allocator.allocate(32)) {
```
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/api/tests/BufferCharSequenceTest.java:35]: Nit: I'd have line breaks between each resource in a twr, like the other existing tests we have.

**Review by chrisvest — APPROVED** [review]

**Pull request merged**